text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,970 @@
+
+> 🏗️ Note for those familiar with the previous RFC 🏗️
+>
+> This is the most critical semantic difference.  Structurally this definition of the trait is very similar to the previous -- there's still a method splitting the type into a discriminated union between two associated types, and constructors to rebuild it from them.  But by keeping the "result-ness" or "option-ness" in the residual type, it gives extra control over interconversion that wasn't possible before.  The changes other than this are comparatively minor, typically either rearrangements to work with that or renamings to change the vocabulary used in the trait.
+
+Using `!` is then just a convenient yet efficient way to create those residual types.  It's nice as a user, too, not to need to understand an additional type.  Just the same "it can't be that one" pattern that's also used in `TryFrom`, where for example `i32::try_from(10_u8)` gives a `Result<i32, !>`, since it's a widening conversion which cannot fail.  Note that there's nothing special going on with `!` here -- any uninhabited `enum` would work fine.
+
+
+## How error conversion works
+
+One thing [The Book mentions](https://doc.rust-lang.org/stable/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator),
+if you recall, is that error values in `?` have `From::from` called on them, to convert from one error type to another.
+
+The previous section actually lied to you slightly: there are *two* traits involved, not just one.  The `from_residual` method is on `FromTryResidual`, which is generic so that the implementation on `Result` can add that extra conversion.  Specifically, the trait looks like this:
+
+```rust
+trait FromTryResidual<Residual = <Self as Try>::Residual> {
+	fn from_residual(r: Residual) -> Self;
+}
+```
+
+And while we're showing code, here's the exact definition of the `Try` trait:
+
+```rust
+trait Try: FromTryResidual {
+	type Output;
+	type Residual;
+	fn branch(self) -> ControlFlow<Self::Residual, Self::Output>;
+	fn from_output(o: Self::Output) -> Self;
+}
+```
+
+The fact that it's a super-trait like that is why I don't feel bad about the slight lie: Every `T: Try` *always* has a `from_residual` function from `T::Residual` to `T`.  It's just that some types might offer more.
+
+Here's how `Result` implements it to do error-conversions:
```

Comment:
```suggestion
Here's how `Result` implements its to do error-conversions:
```

Reply:
I think the current version is correct.

Reply:
"implements its error-conversions" or "implements it to do error-conversions" are correct.

In "implements its error-conversions", "it" refers to `Result`.

In "implements it to do error-conversions", "it" refers to `FromTryResidual`.

For clarity, it seems better not to use "it" here as `FromTryResidual` is separated at least by the digression to show the `Try` trait definition. So instead, how about

```suggestion
Here's how `Result` implements `FromTryResidual` to do error-conversions:
```

Reply:
Sorry, of course it is correct (my brain had problems with 3 digraphs after one another 😄 )